### PR TITLE
Support `condition` as a function

### LINKED ISSUE #1: Handle function `condition` argument
It could be useful to support a function `condition` that is called when source promise settle, to handle constructs like:

``` js
const isEmpty = arr => arr.length === 0;
getList()
    .then(pIf(isEmpty, arr => arr.concat(42))))
    .then(list => {
        console.log(list);
    });
```

### PR DESCRIPTION
**Pull request by sotojuan**
Closes #1.

I just used `is-fn` to make the function act accordingly. If it looks good, I'll do the docs!

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,2 +1,11 @@
 'use strict';
-module.exports = (cond, doIf, doElse) => val => cond ? doIf(val) : (doElse ? doElse(val) : val);
+
+const isFn = require('is-fn');
```
Comment (sindresorhus) [inline index.js]: From the `is-fn` readme:

> In Node.js, just use typeof fn which works exactly the same.

So really no point in this.

File: index.js [hunk 2]
```diff
@@ -1,2 +1,11 @@
...
+
+module.exports = (cond, doIf, doElse) => val => {
+	if (isFn(cond)) {
```
Comment (sindresorhus) [inline index.js]: It should be a function that can also return a promise that is awaited. See how I solved that here: https://github.com/sindresorhus/p-catch-if/blob/454fa343bbf147db4b00c0ba6fb2e103e997e570/index.js#L18-L20

Feel free to steal the tests from here too: https://github.com/sindresorhus/p-catch-if/blob/454fa343bbf147db4b00c0ba6fb2e103e997e570/test.js#L68-L89 (Needs to be adjusted though)

File: test.js [hunk 3]
```diff
@@ -30,3 +30,16 @@ test('composability', async t => {
 
 	t.is(val, 'else');
 });
+
+test('cond can be a function', async t => {
```
Comment (sindresorhus) [inline test.js]: cond => condition

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sotojuan** [comment]
All right, I _think_ I got it. I kept the old tests for a regular function but I can remove those if needed. `cond` now works if it's a function that returns a Promise.

File: index.js [hunk 4]
```diff
@@ -1,2 +1,10 @@
 'use strict';
-module.exports = (cond, doIf, doElse) => val => cond ? doIf(val) : (doElse ? doElse(val) : val);
+
+module.exports = (cond, doIf, doElse) => val => {
+	if (typeof cond === 'function') {
+		return Promise.resolve(cond(val))
+			.then(bool => bool === true ? doIf(val) : (doElse ? doElse(val) : val));
+	}
+
+	return cond ? doIf(val) : (doElse ? doElse(val) : val);
```
Comment (sindresorhus) [inline index.js]: This needs to test for `true` too.

You also have the same logic in two places now, which could be DRY'd up into a function.

File: package.json [hunk 5]
```diff
@@ -40,5 +40,6 @@
   },
   "xo": {
     "esnext": true
-  }
+  },
+  "dependencies": {}
```
Comment (sindresorhus) [inline package.json]: Remove

**Comment by sindresorhus** [comment]
> I kept the old tests for a regular function but I can remove those if needed. 

Keep 'em.

**Comment by sindresorhus** [comment]
Readme needs to be updated.

**Comment by sotojuan** [comment]
🐔

**Comment by sindresorhus** [comment]
There are still some unresolved inline feedback.

File: readme.md [hunk 6]
```diff
@@ -45,9 +45,9 @@ Returns a [thunk](https://en.m.wikipedia.org/wiki/Thunk) that returns a `Promise
 
 #### condition
 
-Type: `boolean`
+Type: `boolean` or `Function`
```
Comment (sindresorhus) [inline readme.md]: Drop the `or`

File: readme.md [hunk 7]
```diff
@@ -45,9 +45,9 @@ Returns a [thunk](https://en.m.wikipedia.org/wiki/Thunk) that returns a `Promise
...
 
-Decides whether `doIf` or `doElse` is executed.
+Decides whether `doIf` or `doElse` is executed. Expected to return a `Promise` that resolves to `true` or the value `true`.
```
Comment (sindresorhus) [inline readme.md]: > Expected to return a `Promise` for `true` or the value `true`.

**Comment by sotojuan** [comment]
Moved over the logic to its own function, but couldn't think of a good name for it. Let me know if I missed anything else 😎

**Review by sindresorhus — APPROVED** [review]

**Comment by sindresorhus** [comment]
Looks great now. Thank you @sotojuan 🤘😎

**Pull request merged**